### DISCUSSION TITLE
Why is string keys so faster than number keys?

### DISCUSSION THREAD

**Discussion by takayama-lily** [nodejs/help#4862] [Q&A]
```js
console.time()
var obj = {
    "1": null, "2": null
}
for (var i = 1e6; i >= 0; i--) {
    Object.keys(obj)
}
console.timeEnd()
// default: 300ms

console.time()
var obj = {
    "a1": null, "a2": null
}
for (var i = 1e6; i >= 0; i--) {
    Object.keys(obj)
}
console.timeEnd()
// default: 30ms
```
`for...in` test got the same result.
I always think number keys should be faster.

**Comment by BridgeAR** [comment]
@nodejs/v8 PTAL

**Reply by verwaest** [reply]
Object.keys and for/in are pretty common operations, so V8 gas a special cache for property names. It's less common for arrays and we generate the array of keys we'll need to visit on-the-fly by calling into the runtime. I'm we could optimize the array case further. /cc @camillobruni

**Reply by camillobruni** [reply]
I filed https://crbug.com/v8/11904 to keep track this. Thanks for reporting!
